Native OpenQASM 3 importer

jakelishman opened the pull request:
### Summary

This is a Qiskit-native version of an OpenQASM 3 parser and converter to `QuantumCircuit`, with the vast majority written in Rust.

At the moment, it exposes two entry points on `qiskit.qasm3`: `loads_experimental` and `load_experimental`, with their API documented.


### Details and comments

This is hugely experimental right now, and this PR is currently missing a few critical features, some because of lack of support in the Rust code here, some because of lack of support on the parser side at https://github.com/Qiskit/openqasm3_parser.

~I don't want to understate that above point: at the moment, `measure` isn't supported (but see Qiskit/openqasm3_parser#42 - it's being _parsed_, just there's a bug in passing it through to the semantic AST).~ Measure and barrier are now both supported, but (for example) `if` statements aren't.

I'm opening this PR in draft now to make it public, but there may be significant changes to the Rust internals, and obviously it will expand with new features from both sides.  This PR currently includes a re-implementation of #11326 (`ExperimentalWarning` - since it was convenient to get it sorted immediately), ~and modifies `qiskit.qasm3.dumps` to use an `OPENQASM 3.0;` header (like #11418, except without the test modifications) because the parser at the moment doesn't like `OPENQASM 3;` (and we should be more accurate anyway).~ now rebased over `main`.

I haven't written the test suite yet.

As a brief example (and tbh, this shows much of the end-to-end supported feature set - it's not huge yet):

```python
import numpy as np
import warnings

import qiskit.qasm2
import qiskit.qasm3
from qiskit.exceptions import ExperimentalWarning
from qiskit.circuit.library import RealAmplitudes

warnings.filterwarnings("ignore", category=ExperimentalWarning)

# This is a little more than 5k gates.
base = RealAmplitudes(27, reps=100, flatten=True)
base = base.assign_parameters(np.random.rand(base.num_parameters))
prog2 = qiskit.qasm2.dumps(base)
prog3 = qiskit.qasm3.dumps(base)

print("OpenQASM 2 time")
%timeit qiskit.qasm2.loads(prog2, custom_instructions=qiskit.qasm2.LEGACY_CUSTOM_INSTRUCTIONS)
print("Current OpenQASM 3 time")
%timeit qiskit.qasm3.loads(prog3)
print("New OpenQASM 3 time")
%timeit qiskit.qasm3.loads_experimental(prog3)
```
```
OpenQASM 2 time
29.1 ms ± 477 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Current OpenQASM 3 time
1.9 s ± 19.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
New OpenQASM 3 time
58.8 ms ± 751 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

jakelishman committed:
Begin work on new OpenQASM 3 importer

jakelishman committed:
Tidy up public interface construction

jakelishman committed:
Add load-from-file entry point

jakelishman committed:
Create public experimental interface

jakelishman committed:
Credit John

While I've been the author of the Qiskit side of this, John wrote the
separate Rust crates that this depends on, so in principle this
contribution to Qiskit is from both of us.

Co-authored-by: John Lapeyre <jlapeyre@users.noreply.github.com>

jakelishman commented on qiskit/qasm/libs/dummy/stdgates.inc:
This dummy version of the standard include is because the base parser can't handle the fancy gate-annotations (`gate cx a, b { ctrl @ x a, b; }` and the like) that the canonical version of the file uses, so this is the same file with all the definitions stubbed out, just for import-parsing purposes.

jlapeyre replied:
`gate cx a, b { ctrl @ x a, b; }` is now parsed correctly by version 0.0.7 of the base parser. (also in 0.0.6,  but into a different structure.)

jakelishman replied:
Ok cool thanks, I'll remove the dummy library if the regular `stdgates.inc` parses.

jakelishman commented on qiskit/exceptions.py:
The changes here are basically a re-implementation of #11326, and they should merge separately to this PR.  I've just included them here as a demonstration of a potential interface for the new parser.

jakelishman commented on crates/qasm3/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+[dependencies]
+pyo3.workspace = true
+hashbrown.version = "0.14.0"
+oq3_semantics = { git = "https://github.com/Qiskit/openqasm3_parser.git", rev = "57407ee" }
```
The base parser isn't released yet, and _technically_, we don't actually need it to be if we specify the dependency as a git revision.  It's kind of like using a submodule, but easier for users.

That said, we totally should release it and fix this dependency before we release Qiskit properly, but for now, this works and is convenient to update.

jakelishman reviewed (commented):
Apologies - the Rust code isn't super well commented or structured at the moment.  I will add more of that in due course.  At the moment, in principle, everything happens in Rust, calling out to Python through PyO3 (so I didn’t need to bother defining a transport format) to construct a `QuantumCircuit`.  The library files are:

* `lib` defines the Python entry points (`loads`)
* `error` at the moment just imports a Python-space exception
* `circuit` defines a few thin Rust-space wrappers around Python objects to inject a little bit more type safety into what I was doing
* `expr` is the constant-folder / expression evaluator (which doesn’t do much constant folding, but it does evaluate a few expressions)
* `build` is where the loop through the ASG happens, and that manages the construction of the `QuantumCircuit`

jakelishman committed:
Fix lint

jlapeyre committed:
Bump version (git hash) of dependency on openqasm3_parser

The qasm3 crate fails to build on Windows. This commit should fix this.

jakelishman committed:
Add support for barrier

jakelishman committed:
Handle measurement with new parser version

Measures, including broadcasted measures, are now supported following
increased support on the parser side. The updated version also includes
changes to a few APIs, and the include-path settings are now
configurable at the entry point, so we switch to those.

jakelishman commented:
Whoopsie, that last commit uses a Rust 1.75 feature: [return-position `impl Trait` in a trait function](https://blog.rust-lang.org/2023/12/21/async-fn-rpit-in-traits.html)!

jakelishman committed:
Fix Rust 1.75 code

jlapeyre committed:
Depend on released version of oq3_seamntics

Change crates/qasm3/Cargo.toml to depend on a released version of the parser.
that is the crate oq3_semantics, rather than a commit of the github repo.
All of the commits added since the commit previously specified in the
dependency are either updating the README or tweaking the github actions.

jlapeyre commented:
NOTE: The following PR has been closed, not merged.
See https://github.com/jakelishman/qiskit-terra/pull/22

jakelishman committed:
Add tests of interface successes

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into qasm3/import

jakelishman commented:
Long overdue removing from draft, sorry about that.

jakelishman committed:
Add dummy library to manifest

mtreinish commented on crates/qasm3/README.md:
```diff
@@ -0,0 +1,4 @@
+# `qiskit._qasm3`
+
+This crate is the Rust-level Qiskit interface to an OpenQASM 3 parser.  The parser itself does not know
+about Qiskit, and this crate interfaces with it in a Qiskit-specific manner to produce `QuantumCircuit`s.
```
Do you want to put any details here about `oq3_semantics`? As opposed to refering to it as an "OpenQASM3 parser" or "the parser".

jakelishman replied:
Oh thanks for the reminder - I wrote this readme stub before the external side even had a name.

jakelishman replied:
Done in acb84d5ef3.

mtreinish commented on crates/qasm3/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+[dependencies]
+pyo3.workspace = true
+indexmap.version = "2.1.0"
+hashbrown.version = "0.14.0"
```
Should we set this at the workspace level too, because I know `accelerate` is using both already and I expect the qasm2 parser is as well.

jakelishman replied:
I think at the time I made workspace versions we didn't actually have the support for inheriting it in the MSRV (I have a vague recollection of defining a lot more in there then stripping it out), but yeah I can pull them up together now.

jakelishman replied:
Done in 9f4e342416.

mtreinish commented on qiskit/qasm3/__init__.py:
```diff
@@ -253,3 +300,25 @@ def loads(program: str):
+
+
+@functools.wraps(_qasm3.loads)
+def loads_experimental(source, /, *, custom_gates=None, include_path=None):
+    """<overridden by functools.wraps>"""
+    warnings.warn(
+        "This is an experimental native version of the OpenQASM 3 importer."
+        " Beware that its interface might change, and it might be missing features.",
+        category=ExperimentalWarning,
+    )
+    return _qasm3.loads(source, custom_gates=custom_gates, include_path=include_path)
+
+
+@functools.wraps(_qasm3.load)
+def load_experimental(pathlike_or_filelike, /, *, custom_gates=None, include_path=None):
```
I wonder instead of a parallel function, if just having a flag `native_parser=True` or something on the existing `load*` functions that emits an `ExperimentalWarning` when set is an easier pattern here. I think either way it's about the same. But at least with the flag, people can use it in place and when we release 1.1 or 1.2 that replaces the python based parser we can just ignore the flag and everything will work the same.

jakelishman replied:
I think a kwarg of the other path might be a bit trickier to get right, because the Rust-only path accepts actual keyword args to control the behaviour and the existing one doesn't.

That said, I'm fine with whatever the decision is on the public interface we expose, I pretty much just need a decision from product (@1ucian0, @kdk).

mtreinish replied:
Ah then in that case then lets just leave it as is (ignore my top level comments to the same effect here) that's a good reason to keep it separate. I was thinking more about the migration path, but we also can just do the same kind of redirect with the separate function.

jlapeyre replied:
I  second @mtreinish s suggestion of a kw arg for the same reasons he gives. If I understand @jakelishman s concern it is that keywords would have to be filtered/forwarded etc. Is that very error prone?

mtreinish commented on Cargo.lock:
```diff
@@ -4,9 +4,9 @@ version = 3
 [[package]]
 name = "ahash"
-version = "0.8.6"
+version = "0.8.7"
```
This is a problem as it will artificially raise our msrv for aarch64 targets (see #11499 for more details)

jakelishman replied:
Yep sorry, that was lazy on my part when I rebuilt the lockfile at one point.  Will fix.

jakelishman replied:
Done in 9f4e342416.

mtreinish commented on crates/qasm3/src/build.rs:
```diff
@@ -0,0 +1,406 @@
+        let name_id = decl
+            .name()
+            .as_ref()
+            .map_err(|err| QASM3ImporterError::new_err(format!("internal error {:?}", err)))?;
```
Isn't the `Display` trait required for an `Error`? I wouldn't expect using `Debug` here to be necessary.

jakelishman replied:
`err` isn't actually an `Error`, it's an `enum oq3_semantics::symbols::SymbolError` which derives `Debug` but not `Display`.

mtreinish replied:
Ah ok, that makes sense then. If we only have debug available then this is correct. I just assumed we had a standard `Error` return from a `Result` here.

jlapeyre replied:
I haven't taken time to think about these things. I suppose `oq3_semantics::symbols::SymbolError` should derive or implement `Display`. I can open a stub of an issue as a reminder.

mtreinish commented on crates/qasm3/src/build.rs:
```diff
@@ -0,0 +1,406 @@
+                            self.add_creg(py, name_id.clone(), name_symbol.name(), *size)
+                        }
+                        _ => Err(QASM3ImporterError::new_err(
+                            "cannot handle quantum registers with more than one dimension",
```
?
```suggestion
                            "cannot handle classical registers with more than one dimension",
```

jakelishman replied:
Done in a2d2c4f3ec.

mtreinish commented on crates/qasm3/src/build.rs:
```diff
@@ -0,0 +1,406 @@
+        call: &asg::GateCall,
+    ) -> PyResult<()> {
+        if call.modifier().is_some() {
+            return Err(QASM3ImporterError::new_err("gate modifiers not handled"));
```
Is this more accurate? I'm assuming this is a current limitation and future releases might have support for gate modifiers:

```suggestion
            return Err(QASM3ImporterError::new_err("gate modifiers currently not handled"));
```

jakelishman replied:
I'll push up a commit handling them, now they're done in the parser.

jakelishman replied:
Done the message in 5b8398ab13, I can do the modifiers separately.

jlapeyre replied:
Note that I started a PR in https://github.com/jakelishman/qiskit-terra/pull/23 . I did not know if you would have time to get to it.

mtreinish commented on crates/qasm3/src/build.rs:
```diff
@@ -0,0 +1,406 @@
+            .as_ref()
+            .map_err(|err| QASM3ImporterError::new_err(format!("internal error {:?}", err)))?;
+        let gate = self.symbols.gates.get(gate_id).ok_or_else(|| {
+            QASM3ImporterError::new_err(format!("internal logic error: unknown gate {:?}", gate_id))
```
Is there a distinction between "internal logic error" and "internal error"?

jakelishman replied:
No, I can make them all consistent.

jakelishman replied:
Oh, actually looking now, there is a bit of a difference - I used "internal error" in situations where I have to handle a case/variant from `oq3_semantics` that _should_ be unreachable, and "internal logic error" if I had to handle a branch that (in theory, at least) would have been an error on this side.  Not quite sure why some places panic rather than raising the correct error, though.

Anyway, I just made them all "internal error" and turned them into proper exceptions in a2d2c4f3ec

mtreinish commented on crates/qasm3/src/build.rs:
```diff
@@ -0,0 +1,406 @@
+                .unwrap_or_default()
+                .iter()
+                .map(|param| expr::eval_gate_param(py, &self.symbols, ast_symbols, param))
+                .collect::<PyResult<Vec<_>>>()?,
```
Is the vec collection required here? Looking at the pytuple new() docs it should take anything that implements `IntoIterator` which includes an existing iterator.

jakelishman replied:
I need to convert the `Iterator<Item = PyResult<T>>` into a `Iterator<Item = T>` before I can give it to `PyTuple`, and I want the result objects to propagate up, so I'm using the impl that lets you collect an iterator of results in a result of vec.  I couldn't immediately think of a safe neat way to do that without an intermediate collection, though maybe you've got one?

mtreinish replied:
Ah yeah I missed that detail, I don't think there is an easy way with `PyTuple` as it doesn't implement `FromIterator` which would let us wrap it in a result like this does.

jlapeyre committed:
Depend on lastest version, 0.0.7, of openqasm3_parser crates

This commit does two things
* Change the dependency on oq3_semantcis from 0.0.5 to 0.0.7

* Change in build.rs to reflect chang in API for gate modifiers from
`gatecall.modifier()` returning `Option<GateModifier>` to
`gatecall.modifiers()` returning `&[GateModifier]`

mtreinish commented on crates/qasm3/src/circuit.rs:
```diff
@@ -0,0 +1,333 @@
+                // _Technically_, allowing access this internal object can let the Rust-space
+                // wrapper get out-of-sync since we keep a direct handle to the list, but in
+                // practice, the field it's viewing is private and "inaccessible" from Python.
+                self.object.clone()
```
```suggestion
            fn to_object(&self, py: Python) -> Py<PyAny> {
                // _Technically_, allowing access this internal object can let the Rust-space
                // wrapper get out-of-sync since we keep a direct handle to the list, but in
                // practice, the field it's viewing is private and "inaccessible" from Python.
                self.object.clone_ref(py)
```

jakelishman replied:
Ha yeah - iirc I wrote this, then about 5 minutes later realised that `Py::clone_ref` exists and that I should have been using it everywhere.  Not surprised I missed a couple of places.

jakelishman replied:
Done in 3345c34f22.

mtreinish reviewed (commented):
Github just started reporting my review window needs to be refreshed for changes made to the PR branch, so I'm leaving my comments so far so I can refresh in the meantime without potentially losing anything. I'll refresh and continue reviewing.

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into qasm3/import

jakelishman committed:
Expand on readme

mtreinish commented on crates/qasm3/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+    include_path: Option<Vec<OsString>>,
+) -> PyResult<circuit::PyCircuit> {
+    let source =
+        if pathlike_or_filelike.is_instance(PyModule::import(py, "io")?.getattr("TextIOBase")?)? {
+            pathlike_or_filelike
+                .call_method0("read")?
+                .extract::<String>()?
+        } else {
+            let path = PyModule::import(py, "os")?
+                .getattr("fspath")?
+                .call1((pathlike_or_filelike,))?
+                .extract::<OsString>()?;
+            ::std::fs::read_to_string(&path).map_err(|err| {
+                QASM3ImporterError::new_err(format!("failed to read file '{:?}': {:?}", &path, err))
+            })?
+        };
```
Looking at this I feel like we could just move the contents here to be pure python, I guess the only advantage of doing it in rust is if you can do `read_to_string()` which will avoid the python overhead and type conversion for file i/o.

jakelishman replied:
the main reason I did it in Rust was because I'd already written everything else in Rust - I wrote the `load` wrapper ages after I wrote the `loads` wrapper, and at the time I hadn't actually written the interface that wrapped them into `qiskit.qasm3`; I was just calling the functions from the Rust extension module directly.

I can pull `load` up to Python space completely, if you like and have it just call the Rust-space `loads`?

mtreinish replied:
This is fine, it doesn't really make a difference. I expect doing it in rust with a file path will be marginally more efficient because you avoid going through a Python string for the file contents.

It was more just when I was reading through and this and all the logic was done via python I was surprised.

jlapeyre replied:
Avoiding conversion overhead for a string containing the OQ3 source makes sense to me. Handling the path name in Python is ok of course.

mtreinish commented on crates/qasm3/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+        ));
+    }
+    let gates = match custom_gates {
+        Some(gates) => gates
+            .into_iter()
+            .map(|gate| (gate.name().to_owned(), gate))
+            .collect(),
+        None => module
+            .getattr(STDGATES_INC_CUSTOM_GATES_ATTR)?
+            .iter()?
+            .map(|obj| {
+                let gate = obj?.extract::<circuit::PyGate>()?;
+                Ok((gate.name().to_owned(), gate))
+            })
+            .collect::<PyResult<HashMap<_, _>>>()?,
```
Is there a model where we want to be able to extend the standard gate definitions with custom ones? Right now it's either or.

jakelishman replied:
Right now it's possible to do - I exposed `qiskit.qasm3.STDGATES_INC_GATES`, so you can copy that and append to it.  Are you thinking of a different interface?

jlapeyre replied:
I'll stump again here for extern gate declarations. 

If you have `extern gate h q;`, then the parser doesn't error when it finds a call to `h`. And the importer knows that it can use its standard definition of `h` . (or whatever qiskit-specific override mechanism exists).

But if you write `gate h q {... contents ..}`, then this unambiguously overrides Qiskit builtins.

jakelishman replied:
John: it's an interestoing language point, but it doesn't affect the issue at hand here, and it's a matter for the OQ3 design committee, not us in our roles on Qiskit.

This attribute _is_ the "qiskit-specific override mechanism" you describe - it doesn't matter how a "gate" symbol appears in the symbol table, this system is for specifying how to relate gate entries to Python-space constructors, with type checking for the signature (that's why the `num_params` and `num_qubits` are given).

mtreinish commented on crates/qasm3/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+            .into_iter()
+            .map(|gate| (gate.name().to_owned(), gate))
+            .collect(),
+        None => module
+            .getattr(STDGATES_INC_CUSTOM_GATES_ATTR)?
+            .iter()?
```
I'm surprised you just didn't do this natively in rust as it's defined below. It feels weird to call out to python to get the attribute which maps back to a definition from the rust side.

jakelishman replied:
The way I wrote it, it pulls the exact same object each time and unwraps down to Rust space, whereas if I'd used the generator from Rust space, I'd have had to reallocate the objects on the Python heap for each call.  Not a big deal, but since I needed to expose the attr to Python space for users anyway, I just had Rust call it directly.  If I'd been writing this function in Python space, I would have set the default of the arg to that tuple, but I can't do that in Rust space.

jakelishman committed:
Centralise dependencies

mtreinish commented on crates/qasm3/src/expr.rs:
```diff
@@ -0,0 +1,387 @@
+            }
+        }
+        ty => Err(QASM3ImporterError::new_err(format!(
+            "expected an angle-like type, but saw {:?}",
```
This might just be my unfamiliarity with oq3, but won't this match an angle type parameter because it's the catch all? With a function name `eval_gate_param` I assumed an angle type input could be passed to it. So you'd get an error message, "expected an angle-like type, but saw Angle"

jakelishman replied:
This function _should_ handle angle types as well, but right now the backend isn't generating anyway, and I'm only handling float literals for simplicity.  There's no angle literal syntax in OQ3, so it's not super easy to generate one (although there's an argument that the `pi` identifier is `angle` typed, but in practice that raises quite a few issues).

mtreinish replied:
I was more concerned with the error message here in case an angle type were encountered, I figured there was a reason for the limitations here. But if that can't happen in practice then there's nothing to worry about here.

jakelishman replied:
Oh good point, I'll put in a specific catch for if the expression type is `angle`.

jakelishman replied:
Done in fcc79dfea2.

mtreinish commented on qiskit/exceptions.py:
```diff
@@ -61,6 +61,25 @@
+    import warnings
+    from qiskit.exceptions import ExperimentalWarning
+
+    warnings.filterwarnings("ignore", category=ExperimentalWarning)
```
Is this still needed anymore now that #11326 has merged?

jakelishman replied:
No, merge commit 0c517fe3d8fe8b3d5187b720cd2c308d5d2f1463 should have fixed it.

mtreinish commented on pyproject.toml:
```diff
@@ -178,6 +178,8 @@ select = [
 extension-pkg-allow-list = [
     "numpy",
     "qiskit._accelerate",
+    "qiskit._qasm2",
```
I'm surprised pylint wasn't complaining about this already.

jakelishman replied:
Looks like I actually tried it before, but it doesn't get all the way:
https://github.com/Qiskit/qiskit/blob/43ea026baba15f3496bef735aa28b5c5801ec3be/qiskit/qasm2/__init__.py#L529-L534

mtreinish reviewed (commented):
I've finished my first pass through this now and overall this looks great! The big open question for me is really on using an experimental flag on the existing parser interface vs a parallel function. I can see the advantages to either, but I'm kinda thinking a kwarg to trigger the new parser on `qiskit.qasm3.load`  is a better approach now.

One place I feel like we need expansion though is the unit tests, there's minimal coverage so far. I think that's fine for this first pass, but we really should expand that in the future, especially with negative test cases to test that we're raising errors correctly and consistently as for a parser that's an important part of the interface.

The only other minor nit  which I would  find annoying in practice using the parser is that the error messages are all lower case. So it'll be `QASM3ImporterError: internal error` where I'd expect `QASM3ImporterError: Internal error`. But that's just me being weird more than anything else.

jakelishman commented:
> The big open question for me is really on using an experimental flag on the existing parser interface vs a parallel function. I can see the advantages to either, but I'm kinda thinking a kwarg to trigger the new parser on qiskit.qasm3.load is a better approach now.

I think a kwarg is a little tricky because the new interface has additional kwargs itself, but I don't feel strongly.  Comment thread about this here: https://github.com/Qiskit/qiskit/pull/11584#discussion_r1473043398

> One place I feel like we need expansion though is the unit tests, there's minimal coverage so far. I think that's fine for this first pass, but we really should expand that in the future, especially with negative test cases to test that we're raising errors correctly and consistently as for a parser that's an important part of the interface.

I wanted to do this for sure, but at the moment this Qiskit side just prints the errors from the parser to stdout rather than me iterating through and collecting them into a neat way to display them Python side.

> The only other minor nit which I would find annoying in practice using the parser is that the error messages are all lower case. So it'll be QASM3ImporterError: internal error where I'd expect QASM3ImporterError: Internal error. But that's just me being weird more than anything else.

I can change this if you prefer - I tend to write simple error messages in lower case since Python does, but I don't feel strongly.  All the OQ2 parser has lower case messages, but I think much of Python-space Qiskit has sentence case.

jakelishman committed:
Fix typing and error typing

jakelishman committed:
Credit Matt

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

jakelishman committed:
Fix clippy

jakelishman committed:
Clone using GIL

jlapeyre commented:
Here is a PR to this branch (to this PR) with some support for gate modifiers. The PR still lacks ctrl and negctrl.
It needs error handling. 

* https://github.com/jakelishman/qiskit-terra/pull/23

jakelishman commented:
John: I'd rather keep this PR in the state it's in right now in terms of features.  The release is tomorrow, so I think it's too late to be adding more here.

jlapeyre commented on crates/qasm3/src/circuit.rs:
```diff
@@ -0,0 +1,333 @@
+        if received_num_params == self.num_params {
+            self.constructor.call1(py, args.as_ref(py))
+        } else {
+            Err(QASM3ImporterError::new_err(format!(
+                "internal error: wrong number of params for {} (got {}, expected {})",
+                &self.name, received_num_params, self.num_params
+            )))
```
These errors should be found by the parser (oq3_semantics). In fact for gate calls, this is already implemented. If we want to check for errors here anyway, it might be a good idea to note in the message that you have found a bug in openqasm3_parser.

jakelishman replied:
They should be, but because it's possible to represent in the typing I have to cover the case here and convert it to a Python exception so it happens loudly, not quietly returning the wrong thing.

jlapeyre replied:
It's possible to represent in typing. But it's possible to design the parser so that the chance that this happens is very small. Including this error handling code is fine. But if triggered, it still has found a bug in the parser.

jakelishman replied:
Well, not necessarily - it could be that this code screwed something up while collecting the parameters or passed the wrong thing to `PyGate::construct`.  I think it's typically best to avoid naming projects in error messages - you never really know how they'll trigger, especially in a year's time.

jlapeyre commented:
> John: I'd rather keep this PR in the state it's in right now in terms of features. The release is tomorrow, so I think it's too late to be adding more here.

Ok. You did say you wanted to push a commit to implement gate modifiers. Did you decide against doing that yourself as well?

jlapeyre commented:
There were several issues with error reporting mentioned, including one I left above. I think errors should be treated as uniformly as possible. And having foreign errors printed with access to their semantics is not ideal. These things should be fixed. But they take a bit of work, and I don't think are a good use of resources for this first PR.

jakelishman commented:
> You did say you wanted to push a commit to implement gate modifiers. Did you decide against doing that yourself as well?

I did, yeah, sorry - I didn't comment on it, but I decided against it pretty soon after I'd written it the first time.

jakelishman commented:
> There were several issues with error reporting mentioned, including one I left above. I think errors should be treated as uniformly as possible. And having foreign errors printed with access to their semantics is not ideal. These things should be fixed. But they take a bit of work, and I don't think are a good use of resources for this first PR.

For its use in Qiskit, this crate needs to be in charge of displaying the errors because it's the entry point.  I didn't look into neater ways to programmatically pass information about the errors up to Python space for the first pass because it looked like it was buried a few stack levels deep in the Rust side, and as you say, it can be done later.  We can definitely use the nice displays from the Rust side in some form, but we'll need to work out the interface of how best to expose it from the Rust side so that this Python side can use it.

Anything marked "internal error:" here _shouldn't_ be reachable and isn't the primary error path, but I have to be safe and convert things to Python exceptions wherever there's a possibility in the typing for it be incorrect.

jakelishman committed:
Format

jakelishman commented:
I've pushed up the formatting fix (I _really_ ought to sort out my editor - I'm so used to having autoformat setup for Python that I always forget I don't have Rust set to format on write), but I've not reverted the `dummy` library directory for `stdgates.inc` because of Qiskit/openqasm3_parser#88.  I don't think that's critical - it's not really a problem to ship with the dummy data directory.

jlapeyre reviewed (dismissed):
This looks ready to go to me.  @mtreinish ?

jakelishman committed:
Improve error for unexpected angle

mtreinish commented on qiskit/qasm3/__init__.py:
```diff
@@ -172,13 +172,60 @@
+subset of OpenQASM 3 it currently supports.
+
+You can use the experimental interface immediately, with similar functions to the main interface
+above:
```
Do you want to mention something about the interface not being considered stable here, and it might change (or be removed) in a future release without any warnings. The experimental api thing is new so it'd be good to highlight this in as many places as we can.

jakelishman replied:
Done in 6e729bfdd7.

mtreinish commented on crates/qasm3/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+/// of the ``stdgates.inc`` custom instructions is located.
+const STDGATES_INC_CUSTOM_GATES_ATTR: &str = "STDGATES_INC_GATES";
+
+/// Load an OpenQASM 3 program from a string into a :class:`.QuantumCircuit`.
+///
```
This needs to have a warning about it being an experimental unstable interface.

jlapeyre replied:
But it should say a bit more than that. It's not really experimental, it's just preliminary. Except for the name, we are committed to the interface.

mtreinish replied:
That's fair, "experimental api" is just the shorthand we've been using to describe this situation. The exact language I put in the stability policy isn't as specific as to what this situation is called, all it says is:

> For the purposes of semantic versioning, the Qiskit public API is considered any documented module, class, function, or method that is not marked as private (with an underscore _ prefix). However, there can be explicit exceptions made for specific documented APIs. In such cases, these APIs will be clearly documented as not being considered stable interfaces yet, and a user-visible warning will be actively emitted on any use of these unstable interfaces.

So we can call it whatever we want as long as we make the unstable nature of the interface explicit in the docstring. (we already have the warning)

jakelishman replied:
What I've written is like this:
```rust
/// .. warning::
///
///     This native version of the OpenQASM 3 importer is currently experimental.  It is typically
///     much faster than :func:`~qiskit.qasm3.loads`, but has a reduced supported feature set,
///     which will expand over time.
```
it's just not pushed because I'm also writing the release note.  Happy to change that if people prefer.

jlapeyre replied:
It's faster and also gives better diagnostics. I imagine the latter might be more useful (But who knows). I don't know if this warning text is the place to say this.
But as long as you have "much faster", I think it's worth saying "much faster, and gives more informative error messages:.

jakelishman replied:
Done in 6e729bfdd7.  I hadn't included the bit about the error messages in line in the individual function docstrings, but it's in the module docs in fd80ea1b6c.

mtreinish commented on crates/qasm3/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+    crate::build::convert_asg(py, result.program(), result.symbol_table(), gates)
+}
+
+/// Load an OpenQASM 3 program from a source file into a :class:`.QuantumCircuit`.
```
This needs to have a warning about it being an experimental unstable interface.

jakelishman replied:
Done in 6e729bfdd7.

mtreinish reviewed (commented):
Thanks for the quick update. One thing I missed in my earlier review is that for an experimental api the requirements are it's clearly documented **and** it raises a warning on use. I just noticed the documentation side of this is missing in the function docstrings. The other missing thing is the release note, I think even as an experimental feature this definitely warrants a feature note. Once those are fixed I think this is good to go.

jakelishman committed:
Add test of gate broadcasting

jakelishman committed:
Add additional warnings about experimental interface

jlapeyre commented on crates/qasm3/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+pyo3.workspace = true
+indexmap.workspace = true
+hashbrown.workspace = true
+oq3_semantics = "0.0.7"
```
Does this pin the version at `0.0.7`?  In principle we could make backward compatible bug fixes in 0.0.8. There are certainly some bugs that would affect this importer.

jlapeyre replied:
It looks like we might want to release `0.1.0` of the parser in order to allow fixes:
https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio

jlapeyre replied:
Then `oq3_semantics = "0.1.0"` would be compatible with `0.1.1`, etc.

jakelishman replied:
Yes, this dependency is equivalent to `>=0.0.7,<0.0.8`. Cargo treats leading zeros specially - it basically treats the leftmost non-zero as the semantic-version "major".  For how we're updating the base library at the moment, I think that's actually fairly accurate (and not a problem on either side).

That said, this doesn't really affect anything, because we build our Rust crates into cdylib Python extensions that we treat give-or-take like fixed binary objects - one of those ways is that we pin our dependencies with the lockfile, so only precise lockfile updates affect the build environment, and (for the most part) are made separately to Rust code changes.  In this case, large tracts of the lockfile are updated because there's so many new dependencies getting pulled in.

jlapeyre replied:
In any case, I just released openqasm3_parser 0.1.0 ., This is the same as 0.0.7. Just changed the version (and reduced the sleep between publishing crates to 60 s) : https://crates.io/crates/oq3_semantics

I'm not sure I follow everything you said about versions. But if I understand correctly, a particular version of Qiskit includes a fixed version of every file, including the `Cargo.lock` files. And these
fix precise versions of Rust dependencies of Qiksit.

jakelishman committed:
Add release note

jakelishman committed:
Add comment about diagnostics in docs

jakelishman commented:
I've done a pass at the release note in c64c8c11ea, though I expect we might expand it or chop/change it as we write the 1.0 prelude next week.

I also threw in an extra test of gate broadcasting, which Coveralls pointed out I'd missed, in 6712183a83.

jakelishman commented:
Just thinking: perhaps the largest thing that makes me nervous about this is the problems around negative float literals.  If we're able to quickly teach the backend lexer to lex float (and int) literals as `-?<whatever-it-is>` and rely on maximal-munch to disambiguate (as we already are in int vs float, really), I'd feel a lot better.

mtreinish commented on qiskit/qasm3/__init__.py:
```diff
@@ -172,13 +172,65 @@
+.. autofunction:: loads_experimental
+
+These two functions are both experimental, meaning they issue an :exc:`.ExperimentalWarning` on
+usage, and their interfaces may be subject to change within the Qiskit 1.0.  In particular, the
```
```suggestion
usage, and their interfaces may be subject to change within the Qiskit 1.x release series.  In particular, the
```

jakelishman replied:
oh lol, I totally meant to write that - it's why it says "within _the_ Qiskit" haha.  Done in f37f279e0a with line wrapping.

mtreinish reviewed (commented):
LGTM now thanks for the update. Just one quick inline comment on some of the new docs.

mtreinish commented:
> Just thinking: perhaps the largest thing that makes me nervous about this is the problems around negative float literals. If we're able to quickly teach the backend lexer to lex float (and int) literals as `-?<whatever-it-is>` and rely on maximal-munch to disambiguate (as we already are in int vs float, really), I'd feel a lot better.

I think it's fine to do this between rc1 and the final release. This api endpoint is explicitly experimental and actively warns as such, so changes in behavior seem fair game at any point. But also since it wouldn't be changing any public interface I'd view this as more of a bugfix than anything else.

jakelishman committed:
Correct release series comment

jakelishman commented:
Yeah, that sounds reasonable to me.

jlapeyre commented:
> perhaps the largest thing that makes me nervous about this is the problems around negative float literals

I spent a lot of time on the weekend and today on gate modifiers, which turns out not to be important for the release! ... Maybe we can call this the fog of major version release. Anyway, I might be able to fix unary minus pretty quickly.

jakelishman commented:
John: sorry about that.  My aim (certainly in the later half) was less to support modifiers end-to-end, but more to allow us to drop the dummy include file for `stdgates.inc`, because that includes modifiers.  The `gphase` thing made that tricky, though.

~For floats: personally I'd rather handle that in the lexer rather than a unary minus op - it means that we don't have to handle constant-folding the unary minus, which means quite a lot of work on this side because there's no real constant-folding machinery.  It's quite common to lex the minus sign as part of the literal.  Ofc in the general case unary minus still needs to work (if not only for things like `- 1.0` which should lex as two separate tokens), but `-1.0` can easily lex to a float literal, then it'll all just work throughout with no changes anywhere else I think.~

*edit*: That above paragraph makes no sense and hallucinates lexer theory, and I think I'm losing my mind.

jlapeyre commented:
> personally I'd rather handle that in the lexer rather than a unary minus op

Yes I agree that seems like the easiest way to handle floats.

jlapeyre commented:
I notice FWIW that the reference lexer for OQ3 does not lex the minus sign as part of the float. There must be some reason people sometimes prefer to lex `-` always as a token by itself. I don't see an advantage really... except that you don't have to account for whitespace in the float literal because its part of lexing all tokens. Also, it's possible to lex `- 1...` (with whitespace) as a single token.

jakelishman commented:
Thanks very much for that!

I'm really sorry - you pointing out we don't do it in the OQ3 ANTLR lexer made me realise why we don't actually lex like that - it'll produce false parse failures for `a-1.2`, for example, which will lex as an identifier followed by a float, which would appear to be an invalid program.

I don't even know where I got the idea that that was common to do lexer side, I'm sorry.

Do you think it's best to stick with this for one release, or switch it back?

jlapeyre commented:
> I don't even know where I got the idea that that was common to do lexer side,

Not a problem. This took only a few minutes. And it's useful to understand the issue.

https://github.com/Qiskit/openqasm3_parser/pull/90 fails to parse `a-1.2` correctly in exactly this way. I don't have tests in place to detect this. Writing tests is laborious and clumsy. The r-a code had a facility to extract tests from comments. I never got around to transferring it to the oq3 parser.


> Do you think it's best to stick with this for one release, or switch it back?

If you mean https://github.com/Qiskit/openqasm3_parser/pull/90, we have to revert. Before this commit,
`rx(a-1.2) q;` is parsed to ASG correctly. Afterwards, it is not. Again, we lack tests. It is relatively easy for me to add tests that simply count semantic errors. This is better than nothing.

I think correcting this between the lexer and the ASG will not be difficult.
